src/Game/index.ts

```diff
@@ -584,14 +599,125 @@ export class Game {
+      const units = this.getTeamsUnits(team);
+      if (!this.state.teamStates[team].researched[resourceType]) {
+        return;
+      }
+      units.forEach((unit) => {
+        if (unit?.type !== Unit.Type.WORKER) {
+          return;
+        }
+        const minable = Game.ALL_DIRECTIONS.map(dir => this.map.getCellByPos(unit.pos.translate(dir))).filter(cell => cell.resource?.type === resourceType && cell.resource.amount > 0);
+        const mineAmount = Math.min(Math.ceil(unit.getCargoSpaceLeft()/minable.length), miningRate);
+        minable.forEach(cell => {
+          if (!reqs.has(cell.pos.toString())) {
+            reqs.set(cell.pos.toString(), []);
+          }
+          const unitCell = this.map.getCellByPos(unit.pos);
+          const req = new ResourceRequest(
+            unit.pos,
+            mineAmount,
+            unitCell.isCityTile() ? undefined : unit as Worker, // should be city tile
+            unitCell.isCityTile() ? this.cities.get(unitCell.citytile.cityid) : undefined,
+          )
+          const hasReq = reqs.get(cell.pos.toString()).some(r => r.equals(req))
+          if (!hasReq) {
+            reqs.get(cell.pos.toString()).push(req);
+          }
+        });
+      });
+    });
+    return reqs;
+  }
+
+  resolveResourceRequests(resourceType: Resource.Types, requests: Map<string, ResourceRequest[]>): void {
+    requests.forEach((reqs: ResourceRequest[], posStr: String) => {
+      const position = Position.fromString(posStr);
+      let amountLeft = this.map.getCell(position.x, position.y).resource.amount;
+      let amountsReqs = reqs.map(r => [r.amount, r]);
+      while(amountsReqs.length > 0 && amountsReqs.map(e => e[0] as number).reduce((a, b) => a + b) > 0 && amountLeft > 0) {
+        // calculate the smallest amount we should fill
+        // should be equal to the lowest request, or the amount that 
+        // mines out the tile
+        // ie if you have three requests [10, 20, 20] fill 10 first
+        const toFill = Math.min(Math.min(...(amountsReqs.map(e => e[0] as number))), Math.floor(amountLeft / amountsReqs.length));
+        amountsReqs.map(e => e[1] as ResourceRequest).forEach(r => {
+          if(r.city) {
+            r.city.fuel += toFill * this.resourceConversionRate(resourceType);
+          } else {
+            const toGive = Math.min(r.worker.getCargoSpaceLeft(), toFill);
+            r.worker.cargo[resourceType] += toGive;
+          }
+        });
+
+        amountsReqs = amountsReqs.map(([amount, req]: [number, ResourceRequest]) => [amount - toFill, req]);
+        amountLeft -= toFill * amountsReqs.length
+        if (amountLeft < amountsReqs.length) {
+          amountLeft = 0;
+        }
+        amountsReqs = amountsReqs.filter(([amount, _]: [number, ResourceRequest]) => amount > 0);
+      }
+      // set the remaining amount to be the new cell total
+      const cell = this.map.getCellByPos(position);
+      cell.resource.amount = amountLeft;
```

Comment:
If amountLeft is 0, should we set cell.resource to undefined?

Reply:
There's some separate part in the code that clears out 0 resource tiles

Reply:
Sounds good